research/conditional/moe_layers/expert_choice.py

```diff
@@ -53,14 +55,27 @@ def __init__(
         self.use_layer_norm = use_layer_norm
         self.expert_inner_function = expert_inner_function
         self.doutput = self.expert_inner_function.doutput
+        self.gate = None
+        self.moe_values_exp = (
+            moe_values_exp
+            if moe_values_exp != -1
+            else torch.nn.Parameter(torch.tensor(1.0))
+        )
 
         assert (
             not self.one_hot_impl or self.group_by_batch
         ), "Not implemented, would require a lot of memory"
         assert not self.softmax_ungrouped or self.group_by_batch
 
-        init = get_init_fun(init_type=init_type, init_scale=init_scale)
-        gate = init((dmodel, n_experts), dmodel)
+        get_gate = init_gate(
+            self,
+            dmodel,
+            get_router_values_from,
+            init_scale,
+            init_type,
+            n_experts,
+            detach_gate,
+        )
```

Comment:
instead of self let's pass inner_experts

Reply:
ok